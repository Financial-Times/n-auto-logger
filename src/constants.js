export const CATEGORIES = {
	FETCH_RESPONSE_OK: 'FETCH_RESPONSE_OK',
	FETCH_RESPONSE_ERROR: 'FETCH_RESPONSE_ERROR',
	FETCH_NETWORK_ERROR: 'FETCH_NETWORK_ERROR',
	NODE_SYSTEM_ERROR: 'NODE_SYSTEM_ERROR',
	CUSTOM_ERROR: 'CUSTOM_ERROR',
};

export const RESULTS = {
	SUCCESS: 'success',
	FAILURE: 'failure',
};

export default {
	CATEGORIES,
	RESULTS,
};
