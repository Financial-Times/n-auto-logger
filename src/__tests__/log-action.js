import logger from '../index';
import logAction from '../log-action';
import { RESULTS } from '../constants';

jest.mock('@financial-times/n-logger');

describe('logAction', () => {
	afterEach(() => {
		jest.resetAllMocks();
	});

	describe('can enhance sync actionFunction', () => {
		it('log actionFunction name as action name', () => {
			const actionFunction = () => null;
			logAction(actionFunction)();
			expect(logger.info.mock.calls[1][0]).toEqual({
				action: 'actionFunction',
				result: RESULTS.SUCCESS,
			});
		});

		it('to invoke actionFunction and return the same result', () => {
			const actionFunction = jest.fn(() => 'foo');
			const param = { test: 'a' };
			const meta = { meta: 'b' };
			const result = logAction(actionFunction)(param, meta);
			expect(actionFunction.mock.calls).toMatchSnapshot();
			const expectedResult = actionFunction(param, meta);
			expect(result).toBe(expectedResult);
		});

		it('to log event when actionFunction success', () => {
			const actionFunction = () => 'foo';
			const param = { test: 'a' };
			const meta = { meta: 'b' };
			logAction(actionFunction)(param, meta);
			expect(logger.info.mock.calls).toMatchSnapshot();
		});

		it('to log actionFunction failure and throw the caught error', () => {
			const errorInstance = { message: 'bar' };
			const actionFunction = () => {
				throw errorInstance;
			};
			const param = { test: 'a' };
			const meta = { meta: 'b' };
			try {
				logAction(actionFunction)(param, meta);
			} catch (e) {
				expect(e).toBe(errorInstance);
				expect(logger.error.mock.calls).toMatchSnapshot();
			}
		});
	});

	describe('can enhance async actionFunction', () => {
		it('log actionFunction name as action name', async () => {
			const actionFunction = async () => null;
			await logAction(actionFunction)();
			expect(logger.info.mock.calls[1][0]).toEqual({
				action: 'actionFunction',
				result: RESULTS.SUCCESS,
			});
		});

		it('to invoke actionFunction and return the same result', async () => {
			const actionFunction = jest.fn(async () => 'foo');
			const param = { test: 'a' };
			const meta = { meta: 'b' };
			const result = await logAction(actionFunction)(param, meta);
			expect(actionFunction.mock.calls).toMatchSnapshot();
			const expectedResult = await actionFunction(param, meta);
			expect(result).toBe(expectedResult);
		});

		it('to log event when actionFunction success', async () => {
			const actionFunction = async () => 'foo';
			const param = { test: 'a' };
			const meta = { meta: 'b' };
			await logAction(actionFunction)(param, meta);
			expect(logger.info.mock.calls).toMatchSnapshot();
		});

		it('to log actionFunction failure and throw the caught error', async () => {
			const errorInstance = { message: 'bar' };
			const actionFunction = async () => {
				throw errorInstance;
			};
			const param = { test: 'a' };
			const meta = { meta: 'b' };
			try {
				await logAction(actionFunction)(param, meta);
			} catch (e) {
				expect(e).toBe(errorInstance);
				expect(logger.error.mock.calls).toMatchSnapshot();
			}
		});
	});

	describe('throw erros if enhancedFunction has invalid input args', () => {
		it('with more than 2 args', () => {
			const actionFunction = () => null;
			const params = { a: 'test' };
			const meta = { b: 'k' };
			const random = 'test';
			const execution = () => logAction(actionFunction)(params, meta, random);
			expect(execution).toThrowErrorMatchingSnapshot();
		});

		it('with non-Object as first arg (param)', () => {
			const callFunction = () => null;
			const params = 'test';
			const meta = { b: 'k' };
			const execution = () => logAction(callFunction)(params, meta);
			expect(execution).toThrowErrorMatchingSnapshot();
		});

		it('with non-Object as second arg (meta)', () => {
			const callFunction = () => null;
			const params = { a: 'foo' };
			const meta = 'bar';
			const execution = () => logAction(callFunction)(params, meta);
			expect(execution).toThrowErrorMatchingSnapshot();
		});
	});
});
