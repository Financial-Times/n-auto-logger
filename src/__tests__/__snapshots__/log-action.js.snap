// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logAction can enhance async actionFunction to invoke actionFunction and return the same result 1`] = `
Array [
  Array [
    Object {
      "test": "a",
    },
    Object {
      "meta": "b",
    },
  ],
]
`;

exports[`logAction can enhance async actionFunction to log actionFunction failure and throw the caught error 1`] = `
Array [
  Array [
    Object {
      "action": "actionFunction",
      "category": "CUSTOM_ERROR",
      "message": "bar",
      "meta": "b",
      "result": "failure",
      "test": "a",
    },
  ],
]
`;

exports[`logAction can enhance async actionFunction to log event when actionFunction success 1`] = `
Array [
  Array [
    Object {
      "action": "actionFunction",
      "meta": "b",
      "test": "a",
    },
  ],
  Array [
    Object {
      "action": "actionFunction",
      "meta": "b",
      "result": "success",
      "test": "a",
    },
  ],
]
`;

exports[`logAction can enhance sync actionFunction to invoke actionFunction and return the same result 1`] = `
Array [
  Array [
    Object {
      "test": "a",
    },
    Object {
      "meta": "b",
    },
  ],
]
`;

exports[`logAction can enhance sync actionFunction to log actionFunction failure and throw the caught error 1`] = `
Array [
  Array [
    Object {
      "action": "actionFunction",
      "category": "CUSTOM_ERROR",
      "message": "bar",
      "meta": "b",
      "result": "failure",
      "test": "a",
    },
  ],
]
`;

exports[`logAction can enhance sync actionFunction to log event when actionFunction success 1`] = `
Array [
  Array [
    Object {
      "action": "actionFunction",
      "meta": "b",
      "test": "a",
    },
  ],
  Array [
    Object {
      "action": "actionFunction",
      "meta": "b",
      "result": "success",
      "test": "a",
    },
  ],
]
`;

exports[`logAction throw erros if enhancedFunction has invalid input args with more than 2 args 1`] = `"invalid input for action function [actionFunction] following signature standard of (param: Object, meta?: Object)"`;

exports[`logAction throw erros if enhancedFunction has invalid input args with non-Object as first arg (param) 1`] = `"invalid input for action function [callFunction] following signature standard of (param: Object, meta?: Object)"`;

exports[`logAction throw erros if enhancedFunction has invalid input args with non-Object as second arg (meta) 1`] = `"invalid input for action function [callFunction] following signature standard of (param: Object, meta?: Object)"`;
